x-pack/platform/plugins/shared/onechat/server/services/tools/builtin/definitions/create_visualization.ts

```diff
@@ -0,0 +1,339 @@
+            selectedChartType = suggestedType;
+            logger.debug(`LLM suggested chart type: ${selectedChartType}`);
+          } else {
+            logger.warn(`LLM suggested invalid chart type: ${suggestedType}, defaulting to metric`);
+          }
+        }
+
+        // Step 2: Get or generate ES|QL query
+        let esqlQuery: string;
+
+        if (!esql) {
+          logger.debug('No ES|QL query or index provided, generating one');
+          const model = await modelProvider.getDefaultModel();
+          const generateEsqlResponse = await generateEsql({
+            nlQuery: existingConfig
+              ? `Existing esql query to modify: "${parsedExistingConfig.dataset.query}"\n\nUser query: ${nlQuery}`
+              : nlQuery,
+            model,
+            esClient: esClient.asCurrentUser,
+          });
+
+          if (!generateEsqlResponse.queries || generateEsqlResponse.queries.length === 0) {
+            throw new Error('Failed to generate ES|QL query');
+          }
+
+          esqlQuery = generateEsqlResponse.queries[0];
+          logger.debug(`Generated ES|QL query: ${esqlQuery}`);
+        } else {
+          // Looks like an ES|QL query
+          esqlQuery = esql;
+          logger.debug('Using provided ES|QL query');
+        }
+
+        // Step 3: Generate visualization configuration using langgraph with validation retry
+        const model = await modelProvider.getDefaultModel();
+        const schema = parse(esqlMetricState.getSchema());
+
+        // Create and invoke the validation retry graph
+        const graph = createVisualizationGraph(model, logger);
+
+        const finalState = await graph.invoke({
+          nlQuery,
+          esqlQuery,
+          chartType: selectedChartType,
+          schema,
+          existingConfig,
+          messages: [],
+          attemptCount: 0,
+          validatedConfig: null,
+          error: null,
+        });
+
+        const { validatedConfig, error, attemptCount } = finalState;
+
+        if (!validatedConfig) {
+          throw new Error(
+            `Failed to generate valid configuration after ${attemptCount} attempts. Last error: ${
+              error || 'Unknown error'
+            }`
+          );
+        }
+
+        return {
+          results: [
+            {
+              type: ToolResultType.visualization,
+              tool_result_id: getToolResultId(),
+              data: {
+                query: nlQuery,
+                visualization: validatedConfig,
+                chartType: selectedChartType,
+                esqlQuery,
+              },
+            },
+          ],
+        };
+      } catch (error) {
+        logger.error(`Error in create_visualization tool: ${error.message}`);
+        return {
+          results: [
+            {
+              type: ToolResultType.error,
+              data: {
+                message: `Failed to create visualization: ${error.message}`,
```

Comment:
We may want to include params like `nlQuery`, `esql`, `chartType`.

```suggestion
                message: `Failed to create visualization: ${error.message}`,
                esql,
                nlQuery,
                chartType
```